CheckServicePrincipal/action.yml

```diff
@@ -26,75 +26,90 @@ runs:
-              $ApplicationsList = (Invoke-WebRequest -Headers $headerParams -Uri $Env:APP_SECRETS -Method GET)
-              $Logs = @()
-              $NextCounter = 0
-              do {
-                  foreach ($event in ($ApplicationsList.Content | ConvertFrom-Json | select -ExpandProperty value)) {
+              $headerParams = @{ 'Authorization' = "$($oauth.token_type) $($oauth.access_token)" }
+              $applicationsList = (Invoke-WebRequest -Headers $headerParams -Uri $env:APP_SECRETS -Method GET)
+              $logs = @()
+              $nextCounter = 0
+              :servicePrincipalSearch do {
+                  foreach ($event in ($applicationsList.Content | ConvertFrom-Json | select -ExpandProperty value)) {
                       $ids = $event.id
-                      $AppName = $event.displayName
-                      $AppID = $event.appId
-                      $secrets = $event.passwordCredentials
-                      $NextCounter++
-                      if ( $AppName -eq $Env:SERVICE_PRINCIPAL ) {
-                           foreach ($s in $secrets) {
-                               $now = Get-Date
-                               $expires = ( New-TimeSpan -Start $now -End $s.endDateTime).ToString("dd")
-                               Write-Output "Is [int]$expires less than or equal to [int]$Env:EXPIRES_WITHIN_DAYS ?"
-                               if ( [int]$expires -le  [int]$Env:EXPIRES_WITHIN_DAYS ){
-                                   Write-Output "Service Principle $Env:SERVICE_PRINCIPAL, $AppName Expires in $expires days"
-                                   $Log = New-Object System.Object
-                                   $Log | Add-Member -MemberType NoteProperty -Name "Application" -Value $AppName
-                                   $Log | Add-Member -MemberType NoteProperty -Name "ExpiresDays" -Value $expires
-                                   $Log | Add-Member -MemberType NoteProperty -Name "Name"        -Value $s.displayName
-                                   $Log | Add-Member -MemberType NoteProperty -Name "StartDate"   -Value $s.startDateTime
-                                   $Log | Add-Member -MemberType NoteProperty -Name "EndDate"     -Value $s.endDateTime                                   
-                                   $Logs += $Log
-                               }
-                           }
+                      $appName = $event.displayName
+                      $appSecrets = $event.passwordCredentials
+                      $nextCounter++
+                      if ( $appName -eq $env:SERVICE_PRINCIPAL ) {
+                          Write-Output "Service Principal matching our search criteria '$($env:SERVICE_PRINCIPAL)' found with $($appSecrets.Count) secret(s)."
+                          $appSecrets = $appSecrets | Sort-Object -Property endDateTime
+                          foreach ( $secret in $appSecrets ) {
```

Comment:
If it's already sorted by expiry date, then we can simply look at the first one and not use a loop

Reply:
Addressed